rclpy/src/rclpy/_rclpy.c

```diff
@@ -4691,6 +4691,98 @@ rclpy_get_node_parameters(PyObject * Py_UNUSED(self), PyObject * args)
   return node_params;
 }
 
+static PyObject *
+rclpy_serialize(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  PyObject * pymsg;
+  if (!PyArg_ParseTuple(args, "O", &pymsg)) {
+    return NULL;
+  }
+
+  // Get type support
+  rosidl_message_type_support_t * ts =
+    (rosidl_message_type_support_t *)rclpy_common_get_type_support(pymsg);
+  if (!ts) {
+    return NULL;
+  }
+
+  destroy_ros_message_signature * destroy_ros_message = NULL;
+  void * ros_msg = rclpy_convert_from_py(pymsg, &destroy_ros_message);
+  if (!ros_msg) {
+    return NULL;
+  }
+
+  // Create a serialized message object
+  rcl_serialized_message_t serialized_msg = rmw_get_zero_initialized_serialized_message();
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+  rcutils_ret_t rcutils_ret = rmw_serialized_message_init(&serialized_msg, 0u, &allocator);
+  if (RCUTILS_RET_OK != rcutils_ret) {
+    destroy_ros_message(ros_msg);
+    PyErr_Format(RCLError,
+      "Failed to initialize serialized message: %s", rcutils_get_error_string().str);
+    return NULL;
+  }
+
+  // Serialize
+  rmw_ret_t rmw_ret = rmw_serialize(ros_msg, ts, &serialized_msg);
+  destroy_ros_message(ros_msg);
+  if (RMW_RET_OK != rmw_ret) {
+    PyErr_Format(RCLError, "Failed to serialize ROS message");
+    rcutils_ret = rmw_serialized_message_fini(&serialized_msg);
+    if (RCUTILS_RET_OK != rcutils_ret) {
+      PyErr_Format(RCLError,
+        "Failed to finalize serialized message: %s", rcutils_get_error_string().str);
+    }
+    return NULL;
+  }
+
+  // Bundle serialized message in a bytes object
+  return Py_BuildValue("y#", serialized_msg.buffer, serialized_msg.buffer_length);
+}
+
+static PyObject *
+rclpy_deserialize(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  const char * serialized_buffer;
+  int serialized_buffer_size;
+  PyObject * pymsg_type;
+  if (!PyArg_ParseTuple(args, "y#O", &serialized_buffer, &serialized_buffer_size, &pymsg_type)) {
+    return NULL;
+  }
+
+  // Get type support
+  rosidl_message_type_support_t * ts =
+    (rosidl_message_type_support_t *)rclpy_common_get_type_support(pymsg_type);
+  if (!ts) {
+    return NULL;
+  }
+
+  // Create a serialized message object
+  rcl_serialized_message_t serialized_msg = rmw_get_zero_initialized_serialized_message();
+  // Just copy pointer to avoid extra allocation and copy
+  serialized_msg.buffer_capacity = serialized_buffer_size;
```

Comment:
can we guarantee that at this point the size and the capacity of the buffer are equivalent?

Reply:
Yes, the input, declare as "y#", is immutable ([Reference](https://docs.python.org/3/c-api/arg.html#strings-and-buffers)).